esp/services/ws_workunits/ws_workunitsService.cpp

```diff
@@ -111,7 +112,38 @@ bool doAction(IEspContext& context, StringArray& wuids, CECLWUActions action, IP
     if (!wuids.length())
         return true;
 
-    Owned<IWorkUnitFactory> factory = getWorkUnitFactory(context.querySecManager(), context.queryUser());
+    if ((action == CECLWUActions_Restore) || (action == CECLWUActions_Archive))
+    {
+        StringBuffer msg;
+        ForEachItemIn(i, wuids)
+        {
+            StringBuffer wuidStr(wuids.item(i));
+            const char* wuid = wuidStr.trim().str();
+            if (isEmpty(wuid))
+            {
+                msg.appendf("Empty Workunit ID at %u. ", i);
+                continue;
+            }
+            if ((action == CECLWUActions_Archive) && !validateWsWorkunitAccess(context, wuid, SecAccess_Full))
+                msg.appendf("Access denied for Workunit %s. ", wuid);
+        }
+        if (!msg.isEmpty())
+            throw makeStringException(ECLWATCH_INVALID_INPUT, msg);
+
+        Owned<ISashaCommand> cmd = archiveOrRestoreWorkunits(wuids, params, action == CECLWUActions_Archive, false);
+        ForEachItemIn(idx, wuids)
+        {
+            StringBuffer reply;
+            cmd->getId(idx, reply);
+
+            const char* wuid = wuids.item(idx);
+            if ((action == CECLWUActions_Restore) && !validateWsWorkunitAccess(context, wuid, SecAccess_Full))
```

Comment:
@afishbeck Yes. we have to restore a WU in order to check for permission. A user may restore the WU, but, cannot view it. When restoring WUs, the existing code throws an exception if permission denied and stops. This PR just adds the permission denied message into the audit log and continues to the next WU.

Reply:
unauthorized users should not be able to restore all workunits really.
As it's an existing issue, I would not tackle in this PR, but a new JIRA should be opened to prevent unauthorized users restoring arbitrary workunits.
@wangkx - can you open a new JIRA and link it to this one?

Reply:
https://track.hpccsystems.com/browse/HPCC-23634